Source/missiles.cpp

```diff
@@ -314,25 +314,25 @@ void CheckMissileCol(Missile &missile, int minDamage, int maxDamage, bool isDama
 
 	bool isMonsterHit = false;
 	const int mid = dMonster[mx][my];
-	if (missile._micaster != TARGET_BOTH && !missile.IsTrap()) {
-		if (missile._micaster == TARGET_MONSTERS) {
-			if (mid != 0 && (mid > 0 || Monsters[abs(mid) - 1]._mmode == MonsterMode::Petrified)) {
+	if (mid != 0) {

```

Comment:
In addition to Fix Fire Ring trap bug, I have made this whole part a bit more readable.
Now, it is a bit clearer without bunch of elses.
It is also now ready to extract function (called when mid != 0).